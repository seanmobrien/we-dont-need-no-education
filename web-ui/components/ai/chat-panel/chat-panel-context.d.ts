/**
 * Type definitions for chat panel context
 * @module components/ai/chat-panel/chat-panel-context
 */
import { createContext } from 'react';
import { ChatPanelContextValue } from './types';
import { ChatPanelProvider } from './chat-panel-provider';
declare module '@/components/ai/chat-panel/chat-panel-context' {
  /**
   * Create the context
   */
  export const ChatPanelContext: ReturnType<
    typeof createContext<ChatPanelContextValue | null>
  >;

  /**
   * Hook to use the chat panel context
   */
  export function useChatPanelContext(options?: {
    required?: boolean;
  }): ChatPanelContextValue;

  export { ChatPanelProvider as default };
}
