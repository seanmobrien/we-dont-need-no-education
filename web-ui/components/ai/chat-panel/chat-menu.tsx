import {
  Divider,
  IconButton,
  ListSubheader,
  Menu,
  MenuItem,
} from '@mui/material';
import { Dispatch, SetStateAction, useCallback, useState } from 'react';
import MoreVertIcon from '@mui/icons-material/MoreVert';

export const ChatMenu = ({
  onResetSession,
  activeModel,
  setActiveModel,
  onFloat,
}: {
  onResetSession?: () => void;
  activeModel: string;
  setActiveModel: Dispatch<SetStateAction<string>>;
  onFloat?: () => void;
}) => {
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const open = Boolean(anchorEl);
  const availableModels = [
    { id: 'lofi', display: 'Paralegal' },
    { id: 'hifi', display: 'Attorney' },
    { id: 'reasoning-medium', display: 'Partner (Medium)' },
    { id: 'reasoning-high', display: 'Partner (High)' },
  ];
  const activeModelDisplayName =
    availableModels.find((x) => x.id === activeModel)?.display || 'Unknown';
  const onMenuClick = useCallback(
    (event: React.MouseEvent<HTMLButtonElement>) => {
      setAnchorEl(event.currentTarget);
    },
    [setAnchorEl],
  );
  const handleClose = useCallback(() => {
    setAnchorEl(null);
  }, [setAnchorEl]);

  const onResetSessionClick = useCallback(() => {
    handleClose();
    onResetSession?.();
  }, [handleClose, onResetSession]);

  const onFloatClick = useCallback(() => {
    handleClose();
    onFloat?.();
  }, [handleClose, onFloat]);

  return (
    <>
      <IconButton edge="end" onClick={onMenuClick} id="chat-menu-button">
        <MoreVertIcon />
      </IconButton>
      <Menu
        id="basic-menu"
        anchorEl={anchorEl}
        open={open}
        onClose={handleClose}
        slotProps={{
          list: {
            'aria-labelledby': 'chat-menu-button',
          },
        }}
      >
        <MenuItem onClick={onResetSessionClick}>Reset chat session</MenuItem>
        <MenuItem onClick={onFloatClick}>Float</MenuItem>
        <Divider />
        <ListSubheader>{`Active Model: ${activeModelDisplayName}`}</ListSubheader>
        {availableModels.map((model) => (
          <MenuItem
            key={model.id}
            onClick={() => {
              setActiveModel(model.id);
              handleClose();
            }}
            selected={model.id === activeModel}
          >
            {model.display}
          </MenuItem>
        ))}
      </Menu>
    </>
  );
};
