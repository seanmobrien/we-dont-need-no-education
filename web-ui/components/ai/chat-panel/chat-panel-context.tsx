'use client';

import { debounce } from "@/lib/react-util/debounce";
import { ChatPanelConfig, ChatPanelContextValue, DockPosition } from './types';
import React, { createContext, useContext, useState, useEffect, useMemo, SetStateAction } from 'react';


/**
 * Local storage keys for persistence
 */
const STORAGE_KEYS = {
  POSITION: 'chatPanelPosition',
  SIZE: 'chatPanelSize',
  DOCK_SIZE: 'chatPanelDockSize',
} as const;
/**
 * Default configuration
 */
const DEFAULT_CONFIG: ChatPanelConfig = {
  position: 'inline',
  size: {
    width: 600,
    height: 500,
  },
  dockSize: 300,
};

/**
 * Load configuration from localStorage
 */
const loadStoredConfig = (): ChatPanelConfig => {
  if (typeof localStorage === 'undefined') {
    return DEFAULT_CONFIG;
  }
  try {
    const position = (localStorage.getItem(STORAGE_KEYS.POSITION) as DockPosition) || DEFAULT_CONFIG.position;
    const storedSize = localStorage.getItem(STORAGE_KEYS.SIZE);
    const storedDockSize = localStorage.getItem(STORAGE_KEYS.DOCK_SIZE);

    const size = storedSize ? JSON.parse(storedSize) : DEFAULT_CONFIG.size;
    const dockSize = storedDockSize ? parseInt(storedDockSize, 10) : DEFAULT_CONFIG.dockSize;

    return {
      position,
      size,
      dockSize,
    };
  } catch (error) {
    console.warn('Failed to load chat panel config from localStorage:', error);
    return DEFAULT_CONFIG;
  }
};

/**
 * Save configuration to localStorage
 */
const saveConfig = (config: ChatPanelConfig): void => {
  if (typeof localStorage === 'undefined') {
    return;
  }

  try {
    localStorage.setItem(STORAGE_KEYS.POSITION, config.position);
    localStorage.setItem(STORAGE_KEYS.SIZE, JSON.stringify(config.size));
    if (config.dockSize !== undefined) {
      localStorage.setItem(STORAGE_KEYS.DOCK_SIZE, config.dockSize.toString());
    }
  } catch (error) {
    console.warn('Failed to save chat panel config to localStorage:', error);
  }
};

/**
 * Create the context
 */
const ChatPanelContext = createContext<ChatPanelContextValue | null>(null);

/**
 * Hook to use the chat panel context
 */
export const useChatPanelContext = (): ChatPanelContextValue => {
  const context = useContext(ChatPanelContext);
  if (!context) {
    throw new Error('useChatPanelContext must be used within a ChatPanelProvider');
  }
  return context;
};

/**
 * Provider component props
 */
export interface ChatPanelProviderProps {
  children: React.ReactNode;
}

/**
 * Chat panel context provider
 */
export const ChatPanelProvider: React.FC<ChatPanelProviderProps> = ({ children }) => {
  const [config, setConfigState] = useState<ChatPanelConfig>(DEFAULT_CONFIG);
  const [isClient, setIsClient] = useState(false);
  const [dockPanel, setDockPanel] = useState<HTMLDivElement | null>(null);

  // Load stored config on client side
  useEffect(() => {
    setIsClient(true);
    setConfig(loadStoredConfig());
  // Empty ref array - we want this to run on the initial mount and that's about it
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  // Sneaky useEffect to support automatically saving config values to local storage whenever they change
  useEffect(() => {
    if (isClient) {
      saveConfig(config);
    }
  }, [config, isClient]);

  const {
    setConfig,
    setPosition,
    setSize,
    setDockSize,
    setFloating,
    isDocked,
    isFloating,
    isInline,
    debounced,
  } = useMemo(() =>{

    const smartSetConfig = (arg: Partial<ChatPanelConfig> | ((prev: ChatPanelConfig) => Partial<ChatPanelConfig>)) => {
      // Get value out of config, then check for object.is equality as a super-quick early-out
      const newConfig: Partial<ChatPanelConfig> = typeof arg === 'function'
        ? arg(config)
        : arg;
      // Guard against premature initialization or reassignment-to-self tomfoolery
      if (!newConfig || !config || Object.is(newConfig, config)) {
        return; // No change, early exit
      }      
      // Guard against unecessary renders by ensuring there's an actual change to be made
      if (
      (newConfig.position && config.position !== newConfig.position)
      || (newConfig.size && (
        config.size.width !== newConfig.size.width
        || config.size.height !== newConfig.size.height
      ))
      || (newConfig.dockSize && config.dockSize !== newConfig.dockSize)      
      )
      {
        // Spread current config and provided partial into a new object to force state update
        setConfigState({
          ...config,
          ...newConfig,
        });
      }
    };
    
    /**
     * Converts config property SetStateAction-style argument into it's ChatPanelConfig equivalent
     * @param key A key identifying the target property.
     * @param value The value to set for the property.
     * @returns A partial config object or a function that takes the previous config and returns a partial config object.
     */
    const wrapConfigProp = <K extends keyof ChatPanelConfig>(
      key: K, 
      value:SetStateAction<Pick<ChatPanelConfig, K>[K]>
    ): Partial<ChatPanelConfig> | ((prev: ChatPanelConfig) => Partial<ChatPanelConfig>) => {
      // Do we have a function property?
      if (typeof value === 'function') {
        return (prev: ChatPanelConfig) => ({
          // yes; call it with the current value and then return it as the new one....smartSetConfig will handle render reduction for us :)
          [key]: value(prev[key]),
        });
      }
      // No, just return the value as a new partial confiog and let smartSetConfig take care of the rest
      return {
        [key]: value,
      };
    };
    const setSizeCallback = (width: number, height: number) => smartSetConfig({ size: { width, height } });
    return {
      setConfig: smartSetConfig,
      setPosition: (position: SetStateAction<DockPosition>) => smartSetConfig(wrapConfigProp('position', position)),
      setSize: setSizeCallback,
      setDockSize: (size: SetStateAction<number | undefined>) => smartSetConfig(wrapConfigProp('dockSize', size)),
      setFloating: (isFloating: SetStateAction<boolean>) => {
        if (typeof isFloating === 'function') {
          smartSetConfig((prev) => ({
            position: isFloating(prev.position === 'floating') ? 'floating' : 'inline',
          }));
        } else {
          smartSetConfig({ position: isFloating ? 'floating' : 'inline' });
        }
      },
      debounced: {
        setSize: debounce<void, (height:number, width: number) => void>(setSizeCallback, 500),
      },
      isDocked: config.position !== 'inline' && config.position !== 'floating',
      isFloating: config.position === 'floating',
      isInline: config.position === 'inline',
    };
  }, [config]);


  // Computed properties
  
  const contextValue: ChatPanelContextValue = {
    config,
    setPosition,
    setSize,
    setDockSize,
    setFloating,
    isDocked,
    isFloating,
    isInline,
    debounced,    
    dockPanel,
    setDockPanel,
  };

  return (
    <ChatPanelContext.Provider value={contextValue}>
      {children}
    </ChatPanelContext.Provider>
  );
};
