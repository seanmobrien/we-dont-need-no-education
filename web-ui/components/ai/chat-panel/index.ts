import ChatPanel from './chat-panel';

export { ChatPanel };
export { ChatPanelProvider, useChatPanelContext } from './chat-panel-context';
export { DockingOverlay, useDocking } from './docking-overlay';
export { DockedPanel } from './docked-panel';
