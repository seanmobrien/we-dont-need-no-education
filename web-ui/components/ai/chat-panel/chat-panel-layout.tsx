'use client';

import React from 'react';
import { Box, styled } from '@mui/material';
import { useChatPanelContext } from './chat-panel-context';

/**
 * Props for the ChatPanelLayout component
 */
export interface ChatPanelLayoutProps {
  children: React.ReactNode;
}

/**
 * Styled component that adjusts layout based on chat panel state
 */
const LayoutContainer = styled(Box, {
  shouldForwardProp: (prop) => !['chatPanelPosition', 'chatPanelSize', 'isDashboardLayout'].includes(prop as string)
})<{ 
  chatPanelPosition: string;
  chatPanelSize: number;
}>(({ chatPanelPosition, chatPanelSize }) => {
  const baseStyles = {
    width: '100%',
    height: '100%',
    transition: 'all 0.3s ease-in-out',
  };



  // For regular layouts, adjust viewport-wide
  switch (chatPanelPosition) {
    case 'left':
      return {
        ...baseStyles,
        paddingLeft: `${chatPanelSize}px`,
      };
    case 'right':
      return {
        ...baseStyles,
        paddingRight: `${chatPanelSize}px`,
      };
    case 'top':
      return {
        ...baseStyles,
        paddingTop: `${chatPanelSize}px`,
      };
    case 'bottom':
      return {
        ...baseStyles,
        paddingBottom: `${chatPanelSize}px`,
      };
    default:
      return baseStyles;
  }
});

/**
 * Layout component that automatically adjusts spacing based on chat panel state
 */
export const ChatPanelLayout: React.FC<ChatPanelLayoutProps> = ({ 
  children, 
}) => {
  
  const { config, setDockPanel } = useChatPanelContext();
  
  // Only apply spacing adjustments for docked positions
  const isDocked = config.position !== 'inline' && config.position !== 'floating';
  const chatPanelSize = isDocked ? (config.dockSize || 300) : 0;

  return (
    <LayoutContainer      
      ref={(node) => setDockPanel(node as HTMLDivElement)}
      chatPanelPosition={config.position}
      chatPanelSize={chatPanelSize}
    >
      {children}
    </LayoutContainer>
  );
};
