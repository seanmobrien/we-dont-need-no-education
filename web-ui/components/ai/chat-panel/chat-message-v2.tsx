import React from 'react';
import { Box, Paper, Typography, Avatar, Stack } from '@mui/material';
import { Message } from 'ai';
import { VirtualItem } from '@tanstack/react-virtual';
import MuiMarkdown from 'mui-markdown';
import ToolInovocation from './tool-invocation';
import { notCryptoSafeKeyHash } from '@/lib/ai/core';

interface ChatMessageV2Props {
  message: {
    parts: Message['parts'];
    role: Message['role'];
    id: string;
    createdAt?: Date;
  };
  virtualRow: VirtualItem;
  onMeasureElement: (node: Element) => void;
}

export const ChatMessageV2: React.FC<ChatMessageV2Props> = ({
  message,
  virtualRow,
  onMeasureElement,
}) => {
  const { parts = [], role, id: messageId, createdAt } = message;
  const isUser = role === 'user';

  return (
    <Box
      key={virtualRow.key}
      ref={(node: Element) => onMeasureElement(node)}
      sx={{
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        transform: `translateY(${virtualRow.start}px)`,
        p: 0.5,
      }}
    >
      <Stack
        direction="row"
        spacing={1}
        alignItems={isUser ? 'flex-end' : 'flex-start'}
      >
        <Box sx={{ width: '100%', textAlign: isUser ? 'right' : 'left' }}>
          {!isUser && (
            <Avatar
              sx={{
                width: 24,
                height: 24,
                float: 'left',
                marginRight: '1em',
                marginTop: '1em',
              }}
            >
              A
            </Avatar>
          )}
          <Paper
            elevation={6}
            sx={{
              p: 2,
              maxWidth: '70%',
              justifyItems: isUser ? 'flex-end' : 'flex-start',
              textAlign: isUser ? 'right' : 'left',
              marginRight: isUser ? 0 : 2,
              marginLeft: isUser ? 'auto' : 0,
              marginBottom: 2,
              display: 'inline-block',
              // bgcolor: isUser ? 'primary.main' : 'grey.200',
              // color: isUser ? 'white' : 'black',
              borderRadius: 2,
            }}
          >
            <Box>
              {parts
                .map((part) => {
                  return part.type === 'text' ? (
                    <MuiMarkdown
                      key={`${messageId}-text-${notCryptoSafeKeyHash(part.text)}`}
                    >
                      {part.text}
                    </MuiMarkdown>
                  ) : part.type === 'tool-invocation' ? (
                    <ToolInovocation
                      key={`${messageId}-tool-${part.toolInvocation.toolCallId}-${part.toolInvocation.state}`}
                      toolInvocation={part.toolInvocation}
                    />
                  ) : null;
                })
                .filter(Boolean)}
            </Box>
            {createdAt && (
              <Typography
                variant="caption"
                sx={{ display: 'block', textAlign: 'right', mt: 0.5 }}
              >
                {createdAt.toLocaleString()}
              </Typography>
            )}
          </Paper>
        </Box>
      </Stack>
    </Box>
  );
};
