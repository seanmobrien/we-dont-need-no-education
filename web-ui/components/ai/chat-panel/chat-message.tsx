import React, { useCallback, useMemo } from 'react';
import { Box, Paper, Typography, Avatar, Stack } from '@mui/material';
import { Message } from 'ai';
import { VirtualItem } from '@tanstack/react-virtual';
import MuiMarkdown from 'mui-markdown';
import ToolInovocation from './tool-invocation';
import { notCryptoSafeKeyHash } from '@/lib/ai/core';

interface ChatMessageV2Props {
  message: {
    parts: Message['parts'];
    role: Message['role'];
    id: string;
    createdAt?: Date;
  };
  virtualRow: VirtualItem;
  addToolResult: <TResult>({}:{ toolCallId: string; result: TResult }) => void;
  onMeasureElement: (node: Element) => void;
}

export const ChatMessageV2: React.FC<ChatMessageV2Props> = ({
  message,
  virtualRow,
  onMeasureElement,
  addToolResult,
}) => {
  const { parts = [], role, id: messageId, createdAt } = message;
  const isUser = role === 'user';

  const measureElementCallback = useCallback((node: Element) => {
    if(node) {
      onMeasureElement(node);
    }
  }, [onMeasureElement]);
  const stableSx = useMemo(() => ({
    container: {
      position: 'absolute',
      top: 0,
      left: 0,
      width: '100%',
      transform: `translateY(${virtualRow.start}px)`,
      p: 0.5,
    },
    stackWrapper: {
      width: '100%', 
      textAlign: isUser ? 'right' : 'left' 
    },
    avatar: {
      width: 24,
      height: 24,
      float: 'left',
      marginRight: '1em',
      marginTop: '1em',
    },
    paper: {
              p: 2,
              maxWidth: '70%',
              justifyItems: isUser ? 'flex-end' : 'flex-start',
              textAlign: isUser ? 'right' : 'left',
              marginRight: isUser ? 0 : 2,
              marginLeft: isUser ? 'auto' : 0,
              marginBottom: 2,
              display: 'inline-block',
              borderRadius: 2,
            },
            dateline: { display: 'block', textAlign: 'right', mt: 0.5 }
  }), [virtualRow.start, isUser])

  return (
    <Box
      key={virtualRow.key}
      ref={measureElementCallback}
      sx={stableSx.container}
      data-index={virtualRow.index}
    >
      <Stack
        direction="row"
        spacing={1}
        alignItems={isUser ? 'flex-end' : 'flex-start'}
      >
        <Box sx={stableSx.stackWrapper}>
          {!isUser && (
            <Avatar
              sx={stableSx.avatar}
            >
              A
            </Avatar>
          )}
          <Paper
            elevation={6}
            sx={stableSx.paper}
          >
            <Box>
              {parts
                .map((part) => {
                  return part.type === 'text' ? (
                    <MuiMarkdown
                      key={`${messageId}-text-${notCryptoSafeKeyHash(part.text)}`}
                    >
                      {part.text}
                    </MuiMarkdown>
                  ) : part.type === 'tool-invocation' ? (
                    <ToolInovocation
                      key={`${messageId}-tool-${part.toolInvocation.toolCallId}-${part.toolInvocation.state}`}
                      toolInvocation={part.toolInvocation}
                      addToolResult={addToolResult}
                    />
                  ) : null;
                })
                .filter(Boolean)}
            </Box>
            {createdAt && (
              <Typography
                variant="caption"
                sx={stableSx.dateline}
              >
                {createdAt.toLocaleString()}
              </Typography>
            )}
          </Paper>
        </Box>
      </Stack>
    </Box>
  );
};
