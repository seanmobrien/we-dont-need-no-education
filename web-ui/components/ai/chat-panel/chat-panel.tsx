'use client';
import React, { useCallback, useEffect, useState } from 'react';
import {
  Box,
  TextField,
  Stack,
  InputAdornment,
  IconButton,
} from '@mui/material';
import PublishIcon from '@mui/icons-material/Publish';
import { useChat } from '@ai-sdk/react';
import { Message, ToolCall } from 'ai';
import Loading from '@/components/general/loading';
import { ChatMenu } from './chat-menu';
import ChatMessage from './chat-message';
import classnames, {
  alignItems,
  display,
  flexDirection,
  width,
} from '@/tailwindcss.classnames';
import {
  AiModelType,
  AnnotatedRetryMessage,
  isAnnotatedRetryMessage,
} from '@/lib/ai/core';
import { log } from '@/lib/logger';
import { generateChatId } from '@/lib/components/ai';
import { enhancedChatFetch } from '@/lib/components/ai/chat-fetch-wrapper';
import { getReactPlugin } from '@/instrument/browser';
import { withAITracking } from '@microsoft/applicationinsights-react-js';

const getThreadStorageKey = (threadId: string): string =>
  `chatMessages-${threadId}`;
const activeThreadStorageKey = 'chatActiveId';

const getInitialThreadId = (): string => {
  if (typeof sessionStorage !== 'undefined') {
    let chatId = sessionStorage.getItem(activeThreadStorageKey);
    if (chatId) {
      return chatId;
    }
    chatId = generateChatId().id;
    sessionStorage.setItem(activeThreadStorageKey, chatId!);
  }
  return generateChatId().id;
};

const loadCurrentMessageState = (): Message[] | undefined => {
  if (typeof localStorage === 'undefined') {
    return undefined;
  }
  const threadId = getInitialThreadId();
  const threadStorageKey = getThreadStorageKey(threadId);
  const messages = localStorage.getItem(threadStorageKey);
  if (!messages) {
    return undefined;
  }
  return JSON.parse(messages) as Array<Message> | undefined;
};

const stable_onFinish = (message: Message) => {
  let threadId: string;
  let messageId = message.id;
  if (messageId) {
    const parts = messageId.split(':');
    if (parts.length < 2) {
      console.warn('warning - unable to extract current thread id');
      messageId = '1';
      threadId = generateChatId().id;
    } else {
      threadId = parts[0];
      messageId = parts[1];
    }
  } else {
    const newId = generateChatId().id;
    messageId = `${newId}:1`;
    threadId = newId;
  }
  if (typeof sessionStorage !== 'undefined') {
    sessionStorage.setItem(activeThreadStorageKey, threadId);
  }
  if (typeof localStorage !== 'undefined') {
    const messages = loadCurrentMessageState();
    const newMessages = [...(messages ?? []), message];
    localStorage.setItem(
      getThreadStorageKey(threadId),
      JSON.stringify(newMessages),
    );
  }
};

const generateChatMessageId = (): string => {
  const threadId = getInitialThreadId();
  const { id: messageId } = generateChatId();
  return `${threadId}:${messageId}`;
};

const ChatPanel = ({ page }: { page: string }) => {
  const [errorMessage, setErrorMessage] = useState<string | null>(null);
  const [threadId, setThreadId] = useState<string>(getInitialThreadId());
  const [initialMessages, setInitialMessages] = useState<Message[] | undefined>(
    undefined,
  );
  const [activeModel, setActiveModel] = useState<string>('hifi');
  const [rateLimitTimeout, setRateLimitTimeout] = useState<
    Map<AiModelType, Date>
  >(new Map<AiModelType, Date>());

  if (!initialMessages) {
    const messages = loadCurrentMessageState();
    if (messages) {
      setInitialMessages(messages);
    }
  }

  const onChatError = useCallback(
    (error: Error) => {
      console.error('Chat error:', error);
      setErrorMessage((current) =>
        current === error.message ? current : error.message,
      );
    },
    [setErrorMessage],
  );
  const onChatToolCall = useCallback(
    async ({ toolCall }: { toolCall: ToolCall<string, unknown> }) => {
      if (toolCall.toolName === 'getLocation') {
        const cities = ['New York', 'Los Angeles', 'Chicago', 'San Francisco'];
        return cities[Math.floor(Math.random() * cities.length)];
      }
    },
    [],
  );

  const onModelTimeout = useCallback(
    ({ data: { model, retryAt } }: AnnotatedRetryMessage) => {
      const now = Date.now();
      const retryAtDate = Date.parse(retryAt);
      if (now < retryAtDate) {
        setRateLimitTimeout((prevTimeout) => {
          const newTimeout = new Map(prevTimeout);
          const retryOn = new Date(retryAtDate);
          const current = newTimeout.get(model);
          if (!current || current.getTime() < retryAtDate) {
            newTimeout.set(model, retryOn);
            return newTimeout;
          }
          return prevTimeout;
        });
      }
    },
    [setRateLimitTimeout],
  );

  const {
    id,
    messages,
    input,
    handleInputChange,
    handleSubmit,
    status,
    data,
    setData,
    reload,
    setMessages,
  } = useChat({
    id: threadId,
    generateId: generateChatMessageId,
    initialMessages,
    maxSteps: 5,
    api: '/api/ai/chat',
    fetch: enhancedChatFetch,
    onToolCall: onChatToolCall,
    onFinish: stable_onFinish,
    onResponse: () => {
      (data ?? []).forEach((item) => {
        if (isAnnotatedRetryMessage(item)) {
          onModelTimeout(item);
        } else {
          log((l) => l.warn('Unhandled item type:', item));
        }
      });
    },
    onError: onChatError,
    experimental_throttle: 100,
  });
  const onResetSession = useCallback(() => {
    sessionStorage.removeItem('chatActiveId');
    setThreadId(generateChatId().id);
    setInitialMessages(undefined);
    setMessages([]);
  }, [setMessages]);
  const onSendClick = useCallback(
    (event: React.MouseEvent<HTMLButtonElement>, model?: AiModelType) => {
      if (id) {
        sessionStorage.setItem('chatActiveId', id);
        if (messages && messages.length) {
          localStorage.setItem(`chatMessages-${id}`, JSON.stringify(messages));
        }
      }
      setErrorMessage(null);
      const withModel = model ?? activeModel;
      handleSubmit(event, {
        allowEmptySubmit: false,
        data: {
          model: withModel,
          page,
          threadId: id,
        },
        headers: {
          'x-active-model': withModel,
          'x-active-page': page,
        },
      });
    },
    [activeModel, handleSubmit, id, messages, page],
  );

  useEffect(() => {
    const timeoutIds: Array<NodeJS.Timeout | number> = [];
    const thisData = [...(data ?? [])].filter(isAnnotatedRetryMessage);
    const onRateLimitTimeout = (model: string) => {
      setData((prevData) => {
        const newData = prevData?.filter(
          (item) => !isAnnotatedRetryMessage(item) || item.data.model !== model,
        );
        return newData?.length === prevData?.length ? prevData : newData;
      });
    };
    if (thisData.length > 0) {
      thisData.forEach(
        ({ data: { retryAt, model } }: AnnotatedRetryMessage) => {
          const thisModel = model;
          const timeout = new Date(Date.parse(retryAt));
          const rateLimitExpires = timeout.getTime() - Date.now();
          if (rateLimitExpires <= 0) {
            onRateLimitTimeout(thisModel);
            reload();
          } else {
            timeoutIds.push(
              setTimeout(() => {
                onRateLimitTimeout(thisModel);
                console.warn('Rate limit timeout expired, resending message.');
                reload();
              }, rateLimitExpires),
            );
          }
        },
      );
      return () => timeoutIds.forEach(clearTimeout);
    }
  }, [rateLimitTimeout, reload, data, setData]);
  if (process.env.IS_BUILDING == '1') {
    console.warn('is building, skipping chat panel rendering');
    return <></>;
  }
  return (
    <Box
      id={`chat-panel-${threadId}`}
      sx={{
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        padding: 2,
        width: '96%',
      }}
    >
      <Stack className="w-full" spacing={2}>
        <TextField
          multiline
          rows={5}
          className="w-full"
          variant="outlined"
          placeholder="Type your message here..."
          value={input}
          onChange={handleInputChange}
          onKeyDown={(e) => {
            if (e.key === 'Enter' && !e.shiftKey && !e.ctrlKey) {
              e.preventDefault();
              onSendClick(e as unknown as React.MouseEvent<HTMLButtonElement>);
            } else if (e.key === 'ArrowUp' && input === '') {
              // If the input is empty, allow the user to navigate through previous messages
              e.preventDefault();
              if (messages && messages.length > 0) {
                for (let i = messages.length - 1; i >= 0; i--) {
                  if (messages[i].role === 'user') {
                    const messageText = messages[i].parts
                      ?.filter((x) => x.type === 'text')
                      .map((x) => x.text)
                      .join(' ')
                      .trim();
                    if (messageText) {
                      handleInputChange({
                        target: { value: messageText },
                      } as React.ChangeEvent<HTMLInputElement>);
                      break;
                    }
                  }
                }
              }
            }
          }}
          sx={{ marginBottom: 2 }}
          InputProps={{
            endAdornment: (
              <InputAdornment position="end">
                <Box
                  sx={{
                    display: 'flex',
                    flexDirection: 'column',
                    alignItems: 'center',
                  }}
                >
                  <IconButton edge="end" onClick={onSendClick}>
                    <PublishIcon />
                  </IconButton>
                  <ChatMenu
                    activeModel={activeModel}
                    setActiveModel={setActiveModel}
                    onResetSession={onResetSession}
                  />
                </Box>
              </InputAdornment>
            ),
          }}
        />
        <Box
          sx={{
            marginTop: 2,
            padding: 2,
            border: '1px solid #ccc',
            borderRadius: 4,
            backgroundColor: '--color-gray-800',
            maxHeight: '450px',
            overflowY: 'auto',
          }}
          className={classnames(
            width('w-full'),
            flexDirection('flex-col'),
            display('flex'),
            alignItems('items-start'),
          )}
        >
          <Loading
            loading={status === 'submitted'}
            errorMessage={errorMessage}
          />
          {messages?.length &&
            messages
              .slice(0, messages.length)
              .reverse()
              .map((message) => (
                <ChatMessage message={message} key={message.id} />
              ))}
        </Box>
      </Stack>
    </Box>
  );
};

export default withAITracking(getReactPlugin(), ChatPanel);
