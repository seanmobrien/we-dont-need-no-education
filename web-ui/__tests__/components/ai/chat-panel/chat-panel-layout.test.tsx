import React from 'react';
import { render } from '@/__tests__/test-utils';
import { ChatPanelLayout } from '@/components/ai/chat-panel/chat-panel-layout';
import { ChatPanelProvider } from '@/components/ai/chat-panel/chat-panel-context';

describe('ChatPanelLayout', () => {
  it('renders with default props snapshot', () => {
    const { container } = render(
      <ChatPanelProvider>
        <ChatPanelLayout>
          <div>Test child content</div>
        </ChatPanelLayout>
      </ChatPanelProvider>,
    );
    expect(container).toMatchSnapshot();
  }, 10000);

  it('renders children correctly', () => {
    const { getByText } = render(
      <ChatPanelProvider>
        <ChatPanelLayout>
          <div>Test child content</div>
        </ChatPanelLayout>
      </ChatPanelProvider>,
    );

    expect(getByText('Test child content')).toBeInTheDocument();
  }, 10000);
});
