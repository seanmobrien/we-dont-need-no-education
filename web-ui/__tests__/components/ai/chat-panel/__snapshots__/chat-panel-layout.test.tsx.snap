// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ChatPanelLayout renders with default props snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-1q5fg2i"
  >
    <div>
      Test child content
    </div>
  </div>
</div>
`;
